### ISSUE TITLE
[2.6.0] `XML_DTD` no longer enables `XML_SetBillionLaughsAttack*` functions in `expat.h`

### ISSUE THREAD

**Issue by davidhall586** [libexpat/libexpat#828]
We've added XML_DTD to our code so we could tune the Billion Laughs Attack settings. When we moved to v 2.6.0, it seems to need XML_GE==1 instead. Perhaps this should be an OR?

**Comment by hartwork** [comment]
Hi @davidhall586, your options with >=2.6.0 are:
- Both disabled
- XML_GE enabled and XML_DTD both enabled
- XML_GE enabled only and XML_DTD disabled

Is my understanding correct that you are asking for option "XML_DTD enabled and XML_GE disabled"? Why would that be beneficial? Have you checked what those options do under the hood? In particular XML_DTD easy easy to misunderstand.

**Comment by hartwork** [comment]
> it seems to need XML_GE==1 instead

@davidhall586 no, you can tune the settings with XML_DTD or with XML_GE or be safe by disabling both. So you gained options (cutting even general entities support away when none was needed if desired), there were none taken away.

**Comment by hartwork** [comment]
@davidhall586 PS: I'd be curious what values you tuned it to and what the nature of your XML input files is.

**Comment by hartwork** [comment]
@davidhall586 it would be great to have an update from you on the matter by Monday evening. Thanks in advance!

**Comment by davidhall586** [comment]
Yes, I tried to do XML_DTD without XML_GE. The reason: we had enabled XML_DTD in v2.5.0 in order to use the billion laugh tuning functions. When we moved to 2.6.0, it stopped including those functions with that setting.

I have no problem switching to XML_GE and have done so. In fact, we'd prefer XML_GE, as we have zero need of any of the other stuff XML_DTD pulls in.

The reason, I raised this ticket is because the documentation seems to say that XML_DTD alone would enable the functions, when it will not in 2.6.0. I had to do some digging to figure out why my stuff stopped working.

So, not for me. I have no problems now, but it might be better (should somebody else run into this edge case) for the documentation match the code.

**Comment by hartwork** [comment]
> Yes, I tried to do XML_DTD without XML_GE. The reason: we had enabled XML_DTD in v2.5.0 in order to use the billion laugh tuning functions. When we moved to 2.6.0, it stopped including those functions with that setting.

@davidhall586 I understand, but the build system and C code reject that combination right away, complaining that XML_DTD requires XML_GE. For proof:

https://github.com/libexpat/libexpat/blob/849da3e3fe727fccef5e96ef35482d66447f06a2/expat/lib/xmlparse.c#L72-L74

https://github.com/libexpat/libexpat/blob/849da3e3fe727fccef5e96ef35482d66447f06a2/expat/CMakeLists.txt#L181-L187

So ending up with an error-free build but without the billion laughs API is just not possible. If you managed to do that, please help me see how and reproduce that case.

> The reason I raised this ticket is because the documentation seems to say that XML_DTD alone would enable the functions, when it will not in 2.6.0. I had to do some digging to figure out why my stuff stopped working.

But it does enable those functions, the documentation is correct. How does it not?

> So, not for me. I have no problems now, but it might be better (should somebody else run into this edge case) for the documentation to match the code.

The documentation is correct OR we still have some misunderstanding.

What do you think?

**Comment by hartwork** [comment]
@davidhall586 PS: I'm starting to think that a reference to commit caa27198637683b15d810737bb8a6a81af19bfa5 could be the most to the point answer here.

**Comment by davidhall586** [comment]
Exactly (caa2719)

Here's the doc that confused me:
https://libexpat.github.io/doc/api/latest/#building

Configuring Expat Using the Pre-Processor

Expat's feature set can be configured using a small number of pre-processor definitions. The definition of this symbols does not affect the set of entry points for Expat, only the behavior of the API and the definition of character types in the case of XML_UNICODE_WCHAR_T. The symbols are:

XML_GE
    Added in Expat 2.6.0. Include support for [general entities](https://www.w3.org/TR/2006/REC-xml-20060816/#sec-physical-struct) (syntax &e1; to reference and syntax <!ENTITY e1 'value1'> (an internal general entity) or <!ENTITY e2 SYSTEM 'file2'> (an external general entity) to declare). With XML_GE enabled, general entities will be replaced by their declared replacement text; for this to work for external general entities, in addition an [XML_ExternalEntityRefHandler](https://libexpat.github.io/doc/api/latest/#XML_SetExternalEntityRefHandler) must be set using [XML_SetExternalEntityRefHandler](https://libexpat.github.io/doc/api/latest/#XML_SetExternalEntityRefHandler). Also, enabling XML_GE makes the functions[ XML_SetBillionLaughsAttackProtectionMaximumAmplification](https://libexpat.github.io/doc/api/latest/#XML_SetBillionLaughsAttackProtectionMaximumAmplification) and[ XML_SetBillionLaughsAttackProtectionActivationThreshold](https://libexpat.github.io/doc/api/latest/#XML_SetBillionLaughsAttackProtectionActivationThreshold) available.
    With XML_GE disabled, Expat has a smaller memory footprint and can be faster, but will not load external general entities and will replace all general entities (except the [predefined five](https://www.w3.org/TR/2006/REC-xml-20060816/#sec-predefined-ent): amp, apos, gt, lt, quot) with a self-reference: for example, referencing an entity e1 via &e1; will be replaced by text &e1;. 
XML_DTD
    Include support for using and reporting DTD-based content. If this is defined, default attribute values from an external DTD subset are reported and attribute value normalization occurs based on the type of attributes defined in the external subset. Without this, Expat has a smaller memory footprint and can be faster, but will not load external parameter entities or process conditional sections. If defined, makes the functions[ XML_SetBillionLaughsAttackProtectionMaximumAmplification](https://libexpat.github.io/doc/api/latest/#XML_SetBillionLaughsAttackProtectionMaximumAmplification) and[ XML_SetBillionLaughsAttackProtectionActivationThreshold](https://libexpat.github.io/doc/api/latest/#XML_SetBillionLaughsAttackProtectionActivationThreshold) available.

It looks like these two act independently and that either would turn on billion laughs functions.

**Comment by hartwork** [comment]
Hi @davidhall586,

> It looks like these two act independently and that either would turn on billion laughs functions.

I agree, and I consider that a feature to some extent. The text about XML_DTD was correct for 2.5.0 and is not "wrong" with 2.6.0 now. I could add "(but XML_GE is already doing that)" in theory but that will arguably make things complex without clear end user benefit. 

Sentence "The definition of this symbols does not affect the set of entry points for Expat, only the behavior of the API and the definition of character types in the case of `XML_UNICODE_WCHAR_T`." should probably be dropped though as inaccurate since it is violated by `XML_ATTR_INFO` (for a long time) and also `XML_GE`. It was also violated by `XML_DTD` in Expat 2.5.0… and that made me just understand that you did in fact uncover a true regression here because the situation for `lib/xmlparse.c` is different than for public header `lib/expat.h`. Awesome! :pray: 

I have created a pull request #836 now to both drop that misleading sentence and fix the public header regression in 2.6.0 via patch…

```diff
diff --git a/expat/lib/expat.h b/expat/lib/expat.h
index 79bbfb61..5a8f5178 100644
--- a/expat/lib/expat.h
+++ b/expat/lib/expat.h
@@ -1042,7 +1042,7 @@ typedef struct {
 XMLPARSEAPI(const XML_Feature *)
 XML_GetFeatureList(void);
 
-#if defined(XML_GE) && XML_GE == 1
+#if defined(XML_DTD) || (defined(XML_GE) && XML_GE == 1)
 /* Added in Expat 2.4.0 for XML_DTD defined and
  * added in Expat 2.6.0 for XML_GE == 1. */
 XMLPARSEAPI(XML_Bool)
```

I'm glad we got to this point in time before the release of Expat 2.6.1, that I planned (and still plan to) do tomorrow in the UTC+1 evening. If anything looks off with #836, now would be the perfect time to bring it to attention.
